### ISSUE TITLE
ViewModels sometimes have a null form controller even though it's set at the Application level

### ISSUE THREAD

**Issue by lognaturel** [getodk/collect#4327]
This is in v1.29.0 manifesting as crashes in `FormEntryViewModel.moveForward` ([Crashlytics](https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/7793b2bc479705b76a05a38b55886a21?time=last-seven-days&versions=v1.29.0%20(4036)&sessionEventKey=5FDC7520002600015A7DDD7BA7D58223_1485970946505974935)), `FormEntryViewModel.moveBackward` ([Crashlytics](https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/8145577b551346217c3bbc8ad9b5392e?time=last-seven-days&versions=v1.29.0%20(4036)&sessionEventKey=5FDC79BB00F6000126ABDAD9E9AE3FE3_1485975707750444807)) and `FormEntryViewModel.openHierarchy` ([Crashlytics](https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/00dbc308babc5ff267adc0e554377112?time=last-seven-days&versions=v1.29.0%20(4036)&sessionEventKey=5FDD1B7703A500011450D7A26E9130A6_1486153783798180265)). I believe I've seen other related logging in other non-fatal issues happening previously but I can't find it at the moment. I remember seeing an attempt to `getAuditEventLogger` from a null form controller.

**Comment by seadowg** [comment]
Reproduction steps for this:

1. Disable background processes in developer settings (thanks to @grzegorz for this discovery)
2. Open any form with a file question (all widgets form for instance) and navigate to it
3. Click “Choose file”
4. Switch to a different app using the task manager
5. Return to Collect
6. Choose a file (you’ll see the form reload)
7. Click “Next”

There might be other cases, but I think if we can confirm with logging that this is the only one we're seeing we'd be in a good place to put in a robust fix.

**Comment by seadowg** [comment]
This non fatal in Firebase (requires auth) will track if the issue here is always due to the `pendingActivityResult` branch: https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/95e40842582ed14bc3912c2225265a17?time=last-seven-days&versions=v1.29.1%20(4040)&sessionEventKey=5FE0858F00C900016FAF092B65270560_1487114325314038165.

We'd expect to see `Calling onActivityResult from loadingComplete` before `java.lang.IllegalStateException: ensureFormController called before formLoaded` in the logs.

**Comment by lognaturel** [comment]
I wrote at https://github.com/getodk/collect/pull/4328#discussion_r546440132 about the `FormController set in App but not ViewModel` log I added that: 

> I think we're going to see this happen after the log for Calling onActivityResult from loadingComplete. Because I think that branch is going to call onActivityResult, then return from loadingComplete and then onResume will be called.

That is happening sometimes but not always. [This](https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/95e40842582ed14bc3912c2225265a17?time=last-seven-days&versions=v1.29.1%20(4040)&sessionEventKey=5FE0783C028600012895044BBDE4E25B_1487100664659713435) is an example that does:
```
02:26:04.967 AM Reloading form and restoring state.
02:26:05.308 AM Deserialization FAILED! Deleting cache file
02:26:05.308 AM java.lang.RuntimeException: FormDef.attachControlsToInstanceData
02:26:06.892 AM Loading instance from savepoint file
02:26:07.226 AM Calling onActivityResult from loadingComplete
02:30:02.106 AM FormController set in App but not ViewModel
02:30:03.560 AM ensureFormController called before formLoaded
```

[This](https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/95e40842582ed14bc3912c2225265a17?time=last-seven-days&versions=v1.29.1%20(4040)&sessionEventKey=5FE075EE02B40001227B58F8DBC962AB_1487097181884679546) is an example that does not:
```
02:16:16.355 AM Reloading form and restoring state.
02:16:18.857 AM Loading instance from savepoint file
02:16:20.855 AM Calling onActivityResult from loadingComplete
02:16:34.098 AM ensureFormController called before formLoaded
```

The `java.lang.RuntimeException: FormDef.attachControlsToInstanceData` is one issue I've been trying to learn more about so interesting that it might be related or at least sometimes co-occur.

I recently added the `Reloading form and restoring state.` log and that is showing us that this issue always originates from the path in `loadForm` where nothing is returned from `getLastCustomNonConfigurationInstance` and we have a `null` `formController`.

Basically I think we've all been honing in on the same kinds of issues.

**Comment by lognaturel** [comment]
Tentative plan for Jan 2021:
- Review https://github.com/getodk/collect/pull/4286/files and make sure it addresses this case
- Get https://github.com/getodk/collect/pull/4286/files to QA
- Once https://github.com/getodk/collect/pull/4286/files has passed QA, look into tests (treating recovery from activity or process death as features)

> the FormController set in App but not ViewModel log

@seadowg's hypothesis is that generally `onResume` will already have been called before we get to `loadingComplete` but sometimes `loadingComplete` might be very fast and `onResume` would happen after.

**Comment by lognaturel** [comment]
Probably a duplicate of #4250 but let's just close both at the same time since this has some useful specific repro steps.
